Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::HashSet

cdumez opened the pull request:
#### 432b319392e8baf556f4cbd4842b426e452f06a1
<pre>
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::HashSet
<a href="https://bugs.webkit.org/show_bug.cgi?id=295643">https://bugs.webkit.org/show_bug.cgi?id=295643</a>

Reviewed by Justin Michaud.

* Source/JavaScriptCore/parser/Parser.h:
(JSC::Scope::collectFreeVariables):
* Source/JavaScriptCore/yarr/YarrParser.h:
(JSC::Yarr::Parser::NamedCaptureGroups::nextAlternative):
(JSC::Yarr::Parser::NamedCaptureGroups::popParenthesis):
* Source/WTF/wtf/HashSet.h:
(WTF::shouldValidateKey&gt;::addAll):
(WTF::shouldValidateKey&gt;::removeAll):
(WTF::shouldValidateKey&gt;::unionWith const):
(WTF::shouldValidateKey&gt;::formUnion): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::queueRemovalsLocked):
(WebCore::AXIsolatedTree::queueNodeUpdate):
* Source/WebCore/animation/BlendingKeyframes.cpp:
(WebCore::BlendingKeyframes::fillImplicitKeyframes):
(WebCore::BlendingKeyframes::analyzeKeyframe):
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::applyKeyframeEffects):
(WebCore::KeyframeEffectStack::allowsAcceleration const):
* Source/WebCore/contentextensions/CombinedURLFilters.cpp:
(WebCore::ContentExtensions::generateSuffixWithReverseSuffixTree):
* Source/WebCore/contentextensions/ImmutableNFANodeBuilder.h:
(WebCore::ContentExtensions::ImmutableNFANodeBuilder::setActions):
* Source/WebCore/contentextensions/NFAToDFA.cpp:
(WebCore::ContentExtensions::extendSetWithClosure):
(WebCore::ContentExtensions::DataConverterWithEpsilonClosure::convert):
(WebCore::ContentExtensions::DataConverterWithEpsilonClosure::extend):
* Source/WebCore/contentextensions/Term.h:
(WebCore::ContentExtensions::Term::generateGraph const):
* Source/WebCore/css/MutableStyleProperties.cpp:
(WebCore::MutableStyleProperties::removeProperties):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
* Source/WebCore/loader/CanvasActivityRecord.cpp:
(WebCore::CanvasActivityRecord::mergeWith):
* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::ElementTargetingController::topologicallySortElements):
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::didReceiveHeaders):
(WebCore::ContentSecurityPolicy::inheritInsecureNavigationRequestsToUpgradeFromOpener):
* Source/WebCore/page/mac/ServicesOverlayController.h:
* Source/WebCore/page/mac/ServicesOverlayController.mm:
(WebCore::ServicesOverlayController::buildPhoneNumberHighlights):
(WebCore::ServicesOverlayController::buildSelectionHighlight):
(WebCore::ServicesOverlayController::replaceHighlightsOfTypePreservingEquivalentHighlights):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::getSupportedTypes):
(WebCore::addToHash):
* Source/WebCore/platform/graphics/cg/UTIRegistry.mm:
(WebCore::setAdditionalSupportedImageTypes):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::mimeTypeCache):
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::setPrevalentDomainsToBlockAndDeleteCookiesFor):
(WebCore::NetworkStorageSession::setPrevalentDomainsToBlockButKeepCookiesFor):
(WebCore::NetworkStorageSession::setDomainsWithUserInteractionAsFirstParty):
* Source/WebCore/style/RuleFeature.cpp:
(WebCore::Style::RuleFeatureSet::add):
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyCustomPropertyImpl):
* Source/WebCore/workers/service/server/SWServerRegistration.cpp:
(WebCore::SWServerRegistration::addCookieChangeSubscriptions):
(WebCore::SWServerRegistration::removeCookieChangeSubscriptions):
* Source/WebKit/GPUProcess/media/RemoteRemoteCommandListenerProxy.cpp:
(WebKit::RemoteRemoteCommandListenerProxy::updateSupportedCommands):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::registerTemporaryBlobFilePaths):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
(WebKit::findActingScrollParent):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::adjustVisibilityForTargetedElements):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
(WebKit::WebExtensionAPIPermissions::verifyRequestedPermissions):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerMIMETypeCache.cpp:
(WebKit::RemoteMediaPlayerMIMETypeCache::addSupportedTypes):
* Tools/TestWebKitAPI/Tests/WTF/HashSet.cpp:
(TestWebKitAPI::TEST(WTF_HashSet, FormUnion)):

Canonical link: <a href="https://commits.webkit.org/297196@main">https://commits.webkit.org/297196@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2b25b534225c8467026320c0d1f518ce7e4afc39

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110776 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30439 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20871 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116805 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61044 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112739 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31119 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39025 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84239 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113724 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24863 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99758 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64680 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24230 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17894 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60598 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/103268 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94256 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17954 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119595 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/109330 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37821 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28112 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93198 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38197 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96025 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93022 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23716 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38054 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15806 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33795 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37714 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43184 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/133605 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37375 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36084 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40713 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39082 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (approved):
I like this much better, nice

darinadler commented on Source/JavaScriptCore/yarr/YarrParser.h:
```diff
@@ -163,13 +163,13 @@ class Parser {
         void nextAlternative()
         {
-            m_nestedCaptureGroupNames.last().formUnion(m_activeCaptureGroupNames.last());
+            m_nestedCaptureGroupNames.last().addAll(m_activeCaptureGroupNames.last());
```
Would it be more efficient to use `std::exchange` here to move and clear?

darinadler commented on Source/JavaScriptCore/yarr/YarrParser.h:
```diff
@@ -183,10 +183,10 @@ class Parser {
             ASSERT(m_nestedCaptureGroupNames.size() > 1);
             ASSERT(m_activeCaptureGroupNames.size() > 1);
-            m_nestedCaptureGroupNames.last().formUnion(m_activeCaptureGroupNames.last());
+            m_nestedCaptureGroupNames.last().addAll(m_activeCaptureGroupNames.last());
```
Consider this?
```suggestion
            m_nestedCaptureGroupNames.last().addAll(WTFMove(m_activeCaptureGroupNames.last()));
```

darinadler commented on Source/JavaScriptCore/yarr/YarrParser.h:
```diff
@@ -183,10 +183,10 @@ class Parser {
 
             // Add all the names seen in this parenthesis to the containing alternative.
-            m_activeCaptureGroupNames[m_activeCaptureGroupNames.size() - 2].formUnion(m_nestedCaptureGroupNames.last());
+            m_activeCaptureGroupNames[m_activeCaptureGroupNames.size() - 2].addAll(m_nestedCaptureGroupNames.last());
```
Consider this?
```suggestion
            m_activeCaptureGroupNames[m_activeCaptureGroupNames.size() - 2].addAll(WTFMove(m_nestedCaptureGroupNames.last()));
```

darinadler commented on Source/WebCore/contentextensions/CombinedURLFilters.cpp:
```diff
@@ -252,7 +252,7 @@ static void generateSuffixWithReverseSuffixTree(NFA& nfa, Vector<ActiveSubtree>&
     if (rootAddResult.isNewEntry) {
         ImmutableCharNFANodeBuilder newNode(nfa);
-        newNode.setActions(actionList.begin(), actionList.end());
+        newNode.setActions(actionList);
```
Consider using `WTFMove` here and making `setActions` take an rvalue reference?

darinadler commented on Source/WebCore/contentextensions/NFAToDFA.cpp:
```diff
@@ -275,7 +275,7 @@ struct DataConverterWithEpsilonClosure {
             result.add(nodeId);
             const UniqueNodeList& nodeClosure = nfaNodeclosures[nodeId];
-            result.add(nodeClosure.begin(), nodeClosure.end());
+            result.addAll(nodeClosure);
```
No need for the local variable.
```suggestion
            result.addAll(nfaNodeclosures[nodeId]);
```

darinadler commented on Source/WebCore/contentextensions/NFAToDFA.cpp:
```diff
@@ -287,7 +287,7 @@ struct DataConverterWithEpsilonClosure {
             if (addResult.isNewEntry) {
                 const UniqueNodeList& nodeClosure = nfaNodeclosures[nodeId];
-                destination.add(nodeClosure.begin(), nodeClosure.end());
+                destination.addAll(nodeClosure);
```
No need for the local variable.
```suggestion
                destination.addAll(nfaNodeclosures[nodeId]);
```

darinadler commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -305,7 +305,7 @@ static constexpr std::array supportedTypesOESTextureHalfFloat {
 };
 
 #define ADD_VALUES_TO_SET(set, array) \
-    set.add(std::begin(array), std::end(array))
+    set.addAll(array)
```
Seems like we can get rid of the macro now

darinadler commented on Source/WebCore/page/ElementTargetingController.cpp:
```diff
@@ -906,7 +906,7 @@ Vector<ElementIdentifier> ElementTargetingController::topologicallySortElements(
 
     const auto elementIDs = elementIDToOccludedElementIDs.keys();
-    unprocessedIDs.add(elementIDs.begin(), elementIDs.end());
+    unprocessedIDs.addAll(elementIDs);
```
No need for the local variable any more.
```suggestion
    unprocessedIDs.addAll(elementIDToOccludedElementIDs.keys());
```

darinadler commented on Source/WebCore/platform/graphics/cg/UTIRegistry.mm:
```diff
@@ -152,7 +152,7 @@ void setAdditionalSupportedImageTypes(const Vector<String>& imageTypes)
         additionalSupportedImageTypes().add(imageType);
         auto mimeTypes = RequiredMIMETypesFromUTI(imageType);
-        MIMETypeRegistry::additionalSupportedImageMIMETypes().add(mimeTypes.begin(), mimeTypes.end());
+        MIMETypeRegistry::additionalSupportedImageMIMETypes().addAll(WTFMove(mimeTypes));
```
No need for local variable.
```suggestion
        MIMETypeRegistry::additionalSupportedImageMIMETypes().addAll(RequiredMIMETypesFromUTI[imageType]);
```

darinadler commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -16438,7 +16438,7 @@ void WebPageProxy::adjustVisibilityForTargetedElements(const Vector<Ref<API::Tar
             info->selectors().map([](auto& selectors) {
                 HashSet<String> result;
                 result.reserveInitialCapacity(selectors.size());
-                result.add(selectors.begin(), selectors.end());
+                result.addAll(selectors);
                 return result;
```
Shouldn't this be using a `HashSet` constructor instead of an `add` function?

cdumez replied:
Yes, I'd have to write such a constructor though.

cdumez committed:
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::HashSet
https://bugs.webkit.org/show_bug.cgi?id=295643

Reviewed by Justin Michaud.

* Source/JavaScriptCore/parser/Parser.h:
(JSC::Scope::collectFreeVariables):
* Source/JavaScriptCore/yarr/YarrParser.h:
(JSC::Yarr::Parser::NamedCaptureGroups::nextAlternative):
(JSC::Yarr::Parser::NamedCaptureGroups::popParenthesis):
* Source/WTF/wtf/HashSet.h:
(WTF::shouldValidateKey>::addAll):
(WTF::shouldValidateKey>::removeAll):
(WTF::shouldValidateKey>::unionWith const):
(WTF::shouldValidateKey>::formUnion): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::queueRemovalsLocked):
(WebCore::AXIsolatedTree::queueNodeUpdate):
* Source/WebCore/animation/BlendingKeyframes.cpp:
(WebCore::BlendingKeyframes::fillImplicitKeyframes):
(WebCore::BlendingKeyframes::analyzeKeyframe):
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::applyKeyframeEffects):
(WebCore::KeyframeEffectStack::allowsAcceleration const):
* Source/WebCore/contentextensions/CombinedURLFilters.cpp:
(WebCore::ContentExtensions::generateSuffixWithReverseSuffixTree):
* Source/WebCore/contentextensions/ImmutableNFANodeBuilder.h:
(WebCore::ContentExtensions::ImmutableNFANodeBuilder::setActions):
* Source/WebCore/contentextensions/NFAToDFA.cpp:
(WebCore::ContentExtensions::extendSetWithClosure):
(WebCore::ContentExtensions::DataConverterWithEpsilonClosure::convert):
(WebCore::ContentExtensions::DataConverterWithEpsilonClosure::extend):
* Source/WebCore/contentextensions/Term.h:
(WebCore::ContentExtensions::Term::generateGraph const):
* Source/WebCore/css/MutableStyleProperties.cpp:
(WebCore::MutableStyleProperties::removeProperties):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
* Source/WebCore/loader/CanvasActivityRecord.cpp:
(WebCore::CanvasActivityRecord::mergeWith):
* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::ElementTargetingController::topologicallySortElements):
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::didReceiveHeaders):
(WebCore::ContentSecurityPolicy::inheritInsecureNavigationRequestsToUpgradeFromOpener):
* Source/WebCore/page/mac/ServicesOverlayController.h:
* Source/WebCore/page/mac/ServicesOverlayController.mm:
(WebCore::ServicesOverlayController::buildPhoneNumberHighlights):
(WebCore::ServicesOverlayController::buildSelectionHighlight):
(WebCore::ServicesOverlayController::replaceHighlightsOfTypePreservingEquivalentHighlights):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::getSupportedTypes):
(WebCore::addToHash):
* Source/WebCore/platform/graphics/cg/UTIRegistry.mm:
(WebCore::setAdditionalSupportedImageTypes):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::mimeTypeCache):
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::setPrevalentDomainsToBlockAndDeleteCookiesFor):
(WebCore::NetworkStorageSession::setPrevalentDomainsToBlockButKeepCookiesFor):
(WebCore::NetworkStorageSession::setDomainsWithUserInteractionAsFirstParty):
* Source/WebCore/style/RuleFeature.cpp:
(WebCore::Style::RuleFeatureSet::add):
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyCustomPropertyImpl):
* Source/WebCore/workers/service/server/SWServerRegistration.cpp:
(WebCore::SWServerRegistration::addCookieChangeSubscriptions):
(WebCore::SWServerRegistration::removeCookieChangeSubscriptions):
* Source/WebKit/GPUProcess/media/RemoteRemoteCommandListenerProxy.cpp:
(WebKit::RemoteRemoteCommandListenerProxy::updateSupportedCommands):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::registerTemporaryBlobFilePaths):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
(WebKit::findActingScrollParent):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::adjustVisibilityForTargetedElements):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm:
(WebKit::WebExtensionAPIPermissions::verifyRequestedPermissions):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerMIMETypeCache.cpp:
(WebKit::RemoteMediaPlayerMIMETypeCache::addSupportedTypes):
* Tools/TestWebKitAPI/Tests/WTF/HashSet.cpp:
(TestWebKitAPI::TEST(WTF_HashSet, FormUnion)):

Canonical link: https://commits.webkit.org/297196@main
